Feature openfoammesh numerola

khiltunen opened the pull request:
This PR implements the missing methods to FoamControlWrapper: add_mesh, get_mesh, delete_mesh and iter_meshes. Interface library for mesh transfer using OpenFoam's objectRegistry is included. Data transfer and control files are implemented taking advantage of PyFoam. Unittests and two working examples to one and two-phase (vof) poiseuille flow are included and a converter example to transfering OpenFoam's mesh to H5CUDS -file.

nathanfranklin commented on foam_controlwrapper/examples/foam_mesh_to_h5cuds.py:
```diff
@@ -0,0 +1,14 @@
+from simphony.io.h5_cuds import H5CUDS
+
+foam_controlwrapper = FoamControlWrapper()
+name = 'poiseuille'
```
`foam`mesh_to_h5cuds.py` is used to generate the H5CUDS-file used in the two examples, right? 

I am trying to run the examples.  Where can i find `poiseuille`?  What are the steps to create `poiseuille.cuds`

khiltunen commented:
Just run the mesh_to_hrcuds.py on example directory. The Foam case directory is 'poiseuille' on example directory. Result h5cuds -file is written also on example directory.

nathanfranklin commented on foam_controlwrapper/tests/test_controlwrapper.py:
```diff
@@ -20,52 +18,117 @@
+        wrapper = FoamControlWrapper()
+        wrapper.add_mesh(self.mesh)
+        mesh_inside_wrapper = wrapper.get_mesh(self.mesh.name)
+        self.assertEquals(self.mesh.name, mesh_inside_wrapper.name)
```
Can additional tests be added to check if the contents of the `self.mesh` and `mesh_inside_wrapper` match up? There are some problems at the moment.  For example:  
```        for face1 in self.mesh.iter_faces():
            face2 = mesh_inside_wrapper.get_face(face1.uid)
            self.assertEquals(face1.uid, face2.uid)
            self.assertEquals(face1.points, face2.points) # this currently fails
```

nathanfranklin replied:
Regarding FoamMesh, I see that edge-related methods and the ``add_xxx`` methods aren't supported at the moment.  For those methods of ABCMesh that are currently implemented, it would be good to test them here or better yet in the file `test_foam_mesh.py`

nathanfranklin replied:
> There are some problems at the moment. For example:
```
for face1 in self.mesh.iter_faces():
   face2 = mesh_inside_wrapper.get_face(face1.uid)
   self.assertEquals(face1.uid, face2.uid)
   self.assertEquals(face1.points, face2.points) # this currently fails
```
I figured out what the problem was.  Will create a PR with the fix and add some related unit testing

nathanfranklin commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+    def __init__(self, name, mesh=None, path=None):
+        self.name = name
+        self.data = dc.DataContainer()
+        self.time = str(0)
```
Maybe 'time' can be switched to '_time' as it is not part of the public interface.

I also see it set here but never updated anywhere.  As it is used to read values from files (in `get_cell` etc), should it be updated at the end of the `FoamControlWrapper`'s `run` method

Can additional tests also be added to check the results before and after a `run` to ensure that things are working right.

khiltunen replied:
'time' can be used to get data for specific time step. Actually there is no plan (or pseudo code) yet how we should treat time dependant data. For example if i have some solution on time step n, how should i receive that data in my application code. For example, what should print code

for cell in mesh_inside_wrapper.iter_cells():
    print cell.data[CUBA.PRESSURE]

With the current implementation you could print nth time step pressure:

mesh_inside_wrapper.time = n
for cell in mesh_inside_wrapper.iter_cells():
    print cell.data[CUBA.PRESSURE]

nathanfranklin replied:
> For example, what should print code

> for cell in mesh_inside_wrapper.iter_cells():
> print cell.data[CUBA.PRESSURE]

The high-level objects provide the **current** state of the modeling engine. So, in your example, the printed values of pressure would correspond to whatever the current time step is. 

For example,  the following prints the results at time steps 0, N and 2N:


```
...
wrapper = openfoam.FoamControlWrapper()
...
wrapper.SP[CUBA.NUMBER_OF_TIME_STEPS] = N
....
mesh_inside_wrapper = wrapper.add_mesh(mesh_from_file)
....
# prints values at timestep 0
for cell in mesh_inside_wrapper.iter_cells():
print cell.data[CUBA.PRESSURE]

wrapper.run()

# prints values at timestep N
for cell in mesh_inside_wrapper.iter_cells():
print cell.data[CUBA.PRESSURE]

wrapper.run()

# prints values at timestep 2N
for cell in mesh_inside_wrapper.iter_cells():
print cell.data[CUBA.PRESSURE]
```

In the current API, the wrapper does not provide historical data.  However our API does provide the user with a way to get historical data by querying the wrapper at different stages that the user is interested in (like in the above example).

khiltunen replied:
This functionality is acheved simply by changing in the end of run -method FoamMesh time to current time step. I would still let the time variable be marked as public, so that we can play around with it until we know how to treat historical data. What do you think about this ?

nathanfranklin replied:
> This functionality is achieved simply by changing in the end of run -method FoamMesh time to current time step

Sounds perfect

> I would still let the time variable be marked as public, so that we can play around with it until we know how to treat historical data. What do you think about this ?

I prefer the single _ (i.e. `_time`) then its clear that its not part of the API (i.e. ABCModelingEngine) but you could still use it (i.e. `wrapper._time`).

> until we know how to treat historical data

Regarding historical data, currently the user can collect whatever historical data he or she needs by querying (and saving etc..) the state as the simulation progresses.  The modeling engine wrapper does not have to maintain the historical data.

To me that seems sufficient but I am not an end user or have an understanding of the historical data related use cases.  If you think, there is a need for all the wrapper to maintain historical data, maybe a discussion should be open by the SSB on the wiki.

khiltunen replied:
I'll make modifications so that time is set in the end of run -method and  that underscore marking is used in FoamMesh for time

itziakos commented on foam_controlwrapper/foam_controlwrapper.py:
```diff
@@ -174,42 +97,179 @@ def run(self):
+                        'Mesh \'{}\` does not exist'.format(
+                            name))
+
+    def read_foammesh(self, name, path):
```
*API suggestion*: The ``read_foammesh`` and ``write_foammesh`` could be defined outside the modeling engine class as module level utility functions.

khiltunen replied:
Could you give me an example of 'module level utility function' ?

itziakos replied:
Define the function at module level instead of a method inside the class. 

```python
def read_foammesh(name, path):
        """Read mesh from OpenFoam case files.

        Parameters
        ----------
        name : str
            name to give to mesh
        path : str
            case directory

        Raises
        ------
        Exception if some mesh fron mesh names list not found
        """
        simphonyfoaminterface.init(name, path)
        simphonyfoaminterface.readMesh(name)
        nPoints = simphonyfoaminterface.getPointCount(name)
        nCells = simphonyfoaminterface.getCellCount(name)
        nFaces = simphonyfoaminterface.getFaceCount(name)
        nEdges = 0

        foamMesh = FoamMesh(name)
        foamMesh.generate_uuidmapping(nPoints, nEdges, nFaces, nCells)
        return foamMesh
```

khiltunen replied:
I changed this to module level function. If i wan't to use it (like in example 'foam_mesh_to_h5cuds.py')  how can i do that. I tried to add to '__init__.py'

from .foam_controlwrapper import FoamControlWrapper, read_foammesh
__all__ = ['FoamControlWrapper', 'read_foammesh']

but it did not worked. Actually could not load openfoam module anymore, so something is broken.

itziakos replied:
If read_foammesh is a module level function then you should be able to do 

```python
from simphony_openfoam.foam_controlwrapper.foam_controlwrapper import read_foammesh
```
or if you included the import in your `__init__` as above you should be able to do
 ```python
from foam_controlwrapper import read_foammesh
```

If there is an ImportError then there should be a traceback. It is commonly a typo or missing imports in the foam-controlwrapper.py.

Let me have a look, I can make an example  PR.

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -80,14 +92,314 @@ def writeFoamFiles(self, caseDirectory, solver):
         for file in fileContent:
             try:
-                f = open(os.path.join(caseDirectory, file), 'w')
+                full_name = os.path.join(caseDirectory, file)
```
Currently if there is an IOError during the `write` operation we end-up with open io file handles. I would suggest to use the ``with`` statement to make sure that the filehandle is properly cleaned up. Furthermore, I do not think that we should capture the `IOError` and re-raise it to a `ValueError`. The IOError exception contains important information for the user/developer that would allow for easier debuging.

suggested code:
```python
for filename in fileContent:
    with os.path.join(caseDirectory, filename) as f:
        f.wrtie(fileContent[file])
```

itziakos replied:
corrected example code

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -80,14 +92,314 @@ def writeFoamFiles(self, caseDirectory, solver):
+                        boundaries[bi+1]['type'] = "empty"
+                        break
+
+        except IOError:
```
Similar, to https://github.com/simphony/simphony-openfoam/pull/10/files#r26569036, please do not change the `IOError` to `ValueError`

itziakos replied:
Please she comment https://github.com/simphony/simphony-openfoam/pull/10/files#r26569208. These is no need for the try...except block.

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -80,14 +92,314 @@ def writeFoamFiles(self, caseDirectory, solver):
+        except IOError:
+            error_str = "File {} does not exist"
+            raise ValueError(error_str.format(boundaryFile))
+        try:
```
similar to https://github.com/simphony/simphony-openfoam/pull/10/files#r26569208

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -80,14 +92,314 @@ def writeFoamFiles(self, caseDirectory, solver):
+            parFile = os.path.join(case, 'constant', 'transportProperties')
+            try:
+                control = ParsedParameterFile(parFile)
+#                control["twoPhase"]["phase1"] = SPExt[CUBAExt.PHASE_LIST][0]
```
commented out code should be removed

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+                + "Cell expected."
+            raise TypeError(error_str)
+
+# this not included for performance reasons
```
I would suggest to include this functionality even if it is not fast at the moment. It will help to have an implementation to check performance improvements against and having proper tests for this functionality using a slow implementation can make sure that future improvements are working properly.

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+            pointCount = foamface.getPointCount(self.name)
+            for label in range(pointCount):
+                point = self.get_point(self._foamPointLabelToUuid[label])
+                yield Point.from_point(point)
```
There is no need to create a new Point instance, since `self.get_point` already does that. 
```python
yield point
```
or 
```python
yield self.get_point(self._foamPointLabelToUuid[label])
```
should be enough

itziakos replied:
I think that the same applies for all the other iter_xxx methods.

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+                point = self.get_point(uid)
+                yield Point.from_point(point)
+
+    def iter_edges(self, edge_uuids=None):
```
It is better to raise `NotImplementedError` since they are not supported,

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+
+        """
+
+# create patch data
```
style: There are a lot of comments that look out of place because of wrong indentation

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+                               self._uuidToFoamLabel[face.uid],
+                               face.points)
+
+    def create_dummy_cellvectordata(self, data_name, dimensionset):
```
API suggestion: These methods (create_xxx..) could also be defined as utility functions to avoid confusion since they are not part of the official api.

itziakos commented on foam_controlwrapper/tests/test_controlwrapper.py:
```diff
@@ -20,52 +18,117 @@
+        mesh_inside_wrapper1 = wrapper.get_mesh(self.mesh.name)
+        mesh_inside_wrapper2 = wrapper.get_mesh(mesh2.name)
+
+        assert(sum(1 for _ in mesh_inside_wrapper1.iter_points()) ==
```
Please use specifc unittest assert methods where possible e.g.here one should use `self.assertEqual`

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+        heading = head.format(version=version, foamclass=foamClass,
+                              location=location, foamobject=foamObject)
+        fileContent = heading + scalarTemplates[solver][data_name]
+        try:
```
please see comment https://github.com/simphony/simphony-openfoam/pull/10/files#r26569036

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+                                        dataName,
+                                        self._uuidToFoamLabel[cell.uid],
+                                        tuple(cell.data[data]), 'vector')
+            elif data == "":
```
In this context `data` is a CUBA key, so it should be not possible to have a value of "". It looks that this check is unnecessary.

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+                + str(cell.uid)
+            raise KeyError(error_str)
+
+        if not isinstance(cell, Cell):
```
We should avoid checking for the Cell type. Other adapters will not necessarily use the native Cell implementation so the `isinstance(cell, Cell)` will return false even though the object passed supports the Cell interface.

khiltunen replied:
removed testing Cell type

itziakos commented:
Please note that this PR is lucking a good number of tests (coverage is only 31.56%).

nathanfranklin commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,772 @@
+                    else:
+                        facePoints.append(
+                            self._uuidToFoamLabel[uuid.UUID(puid,
+                                                            version=4)])
```
the if/else statement could be just replaced with just
```
facePoints.append(self._uuidToFoamLabel[puid])
```

khiltunen replied:
This is a fix to overcome FileMesh bug in face.points and cell.points uids (returned type is numpy .string)

itziakos replied:
Please add a comment in the code pointing to the related github issue when workarounds have to take place because of bugs. However, it is not clear why ``FileMesh`` effects the implementation of the openfoam adapter.

khiltunen replied:
Incosistent representation of UID #98

itziakos replied:
UID issue has been fixed in simphony-commons version 0.1.2 (see https://github.com/simphony/simphony-common/issues/98)

itziakos commented on openfoam-interface/libpythonfoam.cpp:
```diff
@@ -0,0 +1,1714 @@
+#define WM_DP 1
```
Please note that there are any tests, directly testing the interface library.

nathanfranklin commented:
@khiltunen   I figured out the problem with cells (i.e. get_cell is returning a cell with the wrong list of ponits ).  I will create a PR in the morning.

khiltunen commented:
@itziakos could you help me on this, why travis gives this error ? 
https://travis-ci.org/simphony/simphony-openfoam/builds/55921268

nathanfranklin commented:
`tables` is needed as you are using H5File in your testing.  you can add it to `dev_requirements.txt`

khiltunen commented:
Thank's Nathan. I think that i have now made the changes suggested and added more unittests to cover testing of modules. Could you please check and point out if something is missing or not working.

nathanfranklin commented:
Is the the time progression working?  Just made a change to the end of `foam_controlwrapper/examples.poiseuille' so that run occurs twice: 

```
mesh_inside_wrapper = wrapper.add_mesh(mesh_from_file)

print "Case directory ", mesh_inside_wrapper.path

# run returns the latest time
lastTime = wrapper.run()
print "Timesteps taken ", lastTime
lastTime = wrapper.run()
print "Timesteps taken ", lastTime
```

and the results are odd:

```
python poiseuille.py
Mesh name  poiseuille

Default patch type set to empty
Case directory  /tmp/tmpWnjhIQ/poiseuille
Timesteps taken  40
Timesteps taken  40
Closing remaining open files:poiseuille/poiseuille.cuds...done
```
and inconsistent between subsequent uses
````
python poiseuille.py
Mesh name  poiseuille

Default patch type set to empty
Case directory  /tmp/tmpZgQuXi/poiseuille
Timesteps taken  40
Timesteps taken  48
Closing remaining open files:poiseuille/poiseuille.cuds...done
```

nathanfranklin commented on foam_controlwrapper/tests/test_controlwrapper.py:
```diff
@@ -6,66 +6,266 @@
+        for face in self.mesh.iter_faces():
+            face_w = mesh_inside_wrapper.get_face(face.uid)
+            self.assertEqual(face.uid, face_w.uid)
+            self.assertEqual(face.points, face_w.points)
```
should the `data` also be checked?    in the setUp method, the each face has a CUBA.LABEL.

khiltunen replied:
I have added the test and modifield the get_face method to support CUBA.LABEL (boundary faces) data. Boundaries are marked using the CUBA.LABEL data, while we don't have any other mechanism to use at a moment (grouping the faces somehow)

nathanfranklin commented on foam_controlwrapper/tests/test_controlwrapper.py:
```diff
@@ -6,66 +6,266 @@
+
+        for cell in self.mesh.iter_cells():
+            cell_w = mesh_inside_wrapper.get_cell(cell.uid)
+            self.assertEqual(cell.uid, cell_w.uid)
```
a  `cell.points` test should be added

nathanfranklin replied:
in the code, it looks like `cell.data` is not empty; the cell.data has velocity and pressure.  could such values be added to the tested cells here.

khiltunen commented:
Made that same test and got same results after consequent run method calls. Actually it took 40 timesteps first and then 7 when the next run call was made (lastTime was 47). Did that test also running simpleFoam from command line on directory made by the example code "poiseuille.py" and get the same behaviour. It's strange that it is working differently in your testing.

nathanfranklin commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,655 @@
+            pointLabels = foamface.getFacePoints(self.name,
+                                                 self._uuidToFoamLabel[uuid])
+            puids = [self._foamPointLabelToUuid[lbl] for lbl in pointLabels]
+            face = Face(puids, uuid)
```
the faces provided by `iter_face` seem to have `data` with a `LABEL`.   `get_face` don't have a label.    the faces provided by these two methods (`iter_face` and `get_face`) should be the same.

nathanfranklin commented:
> Made that same test and got same results after consequent run method calls. Actually it took 40 timesteps first and then 7 when the next run call was made (lastTime was 47). Did that test also running simpleFoam from command line on directory made by the example code "poiseuille.py" and get the same behaviour. It's strange that it is working differently in your testing.

I have run it a few more times.   and gotten: "40,40" then "0,40" and "0,40" and then "40, 8"

What should be the current  TIME_STEP be after the first run and then second run? With the settings in the example ( `wrapper.SP[CUBA.TIME_STEP] = 1` and `wrapper.SP[CUBA.NUMBER_OF_TIME_STEPS] = 1000`), I would have suspected that it would have run 1000 steps and then another 1000 steps.

nathanfranklin commented on foam_controlwrapper/tests/test_controlwrapper.py:
```diff
@@ -6,66 +6,266 @@
+        mesh_file = H5CUDS.open(os.path.join('foam_controlwrapper',
+                                             'tests',
+                                             'simplemesh.cuds'))
+        mesh_from_file = mesh_file.get_mesh(name)
```
i noticed that the cell's in this mesh have VELOCITY and PRESSURE (in their data).

I got two related questions:
*  Are they given default values by the wrapper?  
*  Does `wrapper.add_mesh` support adding meshs where cell's *do* have these VELOCITY and PRESSURE values

nathanfranklin replied:
> i noticed that the cell's in this mesh have VELOCITY and PRESSURE (in their data).

I mean to write that the cell's in this mesh (from the file) DO NOT have VELOCITY or PRESSURE (in their data).

khiltunen replied:
wrapper.add_mesh is supporting data transfer to cells

khiltunen replied:
Concerning the NUMBER_OF_TIME_STEPS: Now the solver runs until its convergence criteria is fullfilled or the number of time steps is reached. Should we add also control for the convergence criteria ?
I continued also testing that strange behaviour of consequent run method calls thjat you got and got also peculiar behaviour when i call it in the loop. It seems that the pyFoam's ConvergnceRunner is getting mixed somehow. I wrote simple Runner class and it start working.

khiltunen commented:
I have added unittests to interface and added Runner class to replace pyFoam's CovergenceRunner. I think that this is ready for merge. Could you please confirm @itziakos and @nathanfranklin

itziakos commented:
@khiltunen there are a number of comments that have not been adressed for example:

- https://github.com/simphony/simphony-openfoam/pull/10#discussion_r26651558
- https://github.com/simphony/simphony-openfoam/pull/10#discussion_r26569208
- https://github.com/simphony/simphony-openfoam/pull/10#discussion_r26571270

Please have a look and let me know if you are going to adressed them.

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -80,14 +112,444 @@ def writeFoamFiles(self, caseDirectory, solver):
+        fileContent = self.create_file_content(caseDirectory, solver,
+                                               time, writeFields)
         for file in fileContent:
             try:
```
I do not think that the wrtie code needs to be not enclosed inside a try..except block. The original IOError should have all the necessary infromation for the user to undetify what went wrong, we should not need to create our own IOError.

khiltunen replied:
All IOError catch are now removed

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -80,14 +112,444 @@ def writeFoamFiles(self, caseDirectory, solver):
+
+# parse system/controlDict -file in case directory
+        parFile = os.path.join(case, 'system', 'controlDict')
+        try:
```
I do not think that you need to use the try..except blocks here, the original error should contain all the necessary information.

itziakos commented:
@khiltunen there is a main comment regarding the codebase. I see the follwoing pattern quite a lot of times
```python
try:
...
except IOError:
     raise IOError
```

This pattern is unecessary since we are not actually handling the error, we are just reraising it with a custom message, which will overwrite any meaningfull information that the original IOError instance had and at the same time change the traceback to hide where the original error had originated.

It is my suggestion that most (some might be actually legitimate checks, but most of what I have seen is not necessary) of these try..except blocks are not needed and make the code very difficult to review.

itziakos commented:
@khiltunen, I will need to go now, I will have a look at the test code tomorrow morning.

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -73,21 +105,346 @@ def writeFoamFiles(self, caseDirectory, solver):
+
+        pressureBCs = BC[CUBA.PRESSURE]
+
+# parse startTime/p -file in case directory
```
:style: mis-aligned comment

itziakos commented on foam_controlwrapper/foam_files.py:
```diff
@@ -1,61 +1,91 @@
 
 
 class FoamFiles():
+    """ Module for OpenFOAM file management
```
In python unlike other languages like functions can exist outside a class. The FoamFiles class does not look like it holds any state and thus the methods in this class do not really need to be part of a class. I would suggest to remove th class boilerplate and define the methods as normal module level functions.

Generic note about python classes:
- New style vs old style classes, `class FoamFiles():` creates an old style class which is not advised, a more appropriate definition would `class FoamFiles(object):`. The difference between the two definitions are subtle but cann effect operation is unexpected ways.
- If the `__init__` does not do anything there is no reason do define an empty `__init__.py`.  There are some exceptions where it is necessary to override the default constructor but it is not the case here.

itziakos replied:
I will add an example alogn the read_foammesh example

itziakos replied:
more information in the `new style`, `old style` issue can be found here http://stackoverflow.com/questions/2715186/why-would-it-be-necessary-to-subclass-from-object-in-python

itziakos commented:
@khiltunen, I have pushed PR #18 on top of this branch with examples of using module level functions instead of enclosing them inside class definitions. Please have a look and let me know if you have any questions.

I will start looking at the new tests now

itziakos commented on foam_controlwrapper/foam_controlwrapper.py:
```diff
@@ -1,36 +1,37 @@
 
     def __init__(self):
 
+        self._meshes = {}
```
When subclassing it is good practice to call the base class implementation (especially for constructors). 
In this case it should be:
```python
def __init__(self):
      super(FoamControlWrapper, self).__init__()
      ....
```

khiltunen replied:
Added that and the same for FoamMesh which was inherited from ABCMesh

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,674 @@
+            dataNames = foamFiles.get_cell_data_names(self.path,
+                                                      self._time)
+            for dataName in dataNames:
+                if dataName == "p":
```
suggestion: There is a good amound of code duplication in this if..elif..else block. In this case you can use a python dictionary to simplify the code.
e.g:
```python
dataName2CUBA = {
    'p' : CUBA.PRESSURE.
    'p_rgh': CUBA.PRESSURE,
    'U': CUBA.VELOCITY,
    'alpha': CUBA.VOLUME_FRACTION,
    'phi': None,
    '': None
}
try:
    cuba = dataName2CUBA[dataName]
except KeyError
    error_str = "Data named "+dataName+" not supported"
    raise NotImplementedError(error_str)
else:
    if cuba is not None:
        cell.data[cuba] = foamFiles.get_cell_data(self.path, self._time, dataName, self._uuidToFoamLabel[uuid])

```

Please note that this is just suggestions and does not have to take place. However, I think that it makes the code easier to read and maintain especially when more dataNames are supported.

itziakos commented on foam_controlwrapper/foam_mesh.py:
```diff
@@ -0,0 +1,674 @@
+                                                  self._time)
+
+        for data in cell.data:
+            if data == CUBA.PRESSURE:
```
Looking at the if else code it makes me wonder that one could refactor the implementation as suggested for https://github.com/simphony/simphony-openfoam/pull/10/files#r27655417. But the wayforward is not trivial. I would howewver make a big difference in how easy it is to maintain the implementation when more CUBA attributes are supported. So this shoudl be somthing for the future.

itziakos commented on foam_controlwrapper/tests/test_controlwrapper.py:
```diff
@@ -6,65 +6,306 @@
+    def tearDown(self):
+        FoamFiles().remove_parser_files(os.getcwd())
+
+    def test_add_mesh(self):
```
The test_add_mesh and test_delete_mesh do not test the specifci functionalit of add_mesh or delete_mesh. They just verify that running the related functions does not raise Execptions.

For example a method defined as follows:
```python
def add_mesh(self, mesh):
      pass

```
would still pass the tests. In short the `test_add_mesh` can **never** fail can only produce an **error**.  A more appropriate name for this method would be `test_add_mesh_does_not_raise_execeptions` but this would also be uncessary as explained below. 

A proper `test_add_mesh` should be based on calling `add_mesh` and verify the calling that method had some effect. In this case (if I am correct) `add_mesh` should create a set of files in a predefined directory. A simple check then would be to **assert** that these tests have been created and have non-zero size. A more advanced check would be to verify the content of these files (but this I understand is not always trivial). To conclude, given that `test_add_mesh` verifies the effects of calling `add_mesh`, the unittest framework will signify a **failure** when assertion failed (e.g. the created files have the wrong names), but will also singify an **error**  if for some reason the test could not complete due to an unknown exception (thus a `test_add_mesh_does_not_raise_execeptions` whould be unecessary).

itziakos replied:
To clarify, executing code and running workflows to detect if anything breaks is still a valid tests methodoligy (i.e. *smoke testing* ) but that check mainly that the code runs not that the code runs **correctly** and thus it is mainly used for higher level integration tests.

itziakos commented on foam_controlwrapper/tests/test_foam_files.py:
```diff
@@ -0,0 +1,100 @@
+import shutil
+
+
+class FoamFilesTestCase(unittest.TestCase):
```
The included test methods are not verify the expected behaviour of the code (please see comments in https://github.com/simphony/simphony-openfoam/pull/10/files#r27658560)

itziakos commented on foam_controlwrapper/tests/test_foam_interface.py:
```diff
@@ -0,0 +1,292 @@
+            coords = foamface.getPointCoordinates(
+                self.name,
+                self._uuidToFoamLabel[point.uid])
+            self.assertEqual(point.coordinates, tuple(coords))
```
I think in the current implementation of the openfoam wrapper Points should have:
```python 
self.assertEqual(point.data, DataContainer())
self.assertiInstance(point.data, DataContainer). 
```

similar for the faces and edges, if I am correct.

itziakos replied:
oops wrong file, please, ingnore my comment

itziakos commented on foam_controlwrapper/tests/test_foam_interface.py:
```diff
@@ -0,0 +1,292 @@
+        if os.path.exists(self.test_dir):
+            shutil.rmtree(self.test_dir)
+
+    def test_add_mesh(self):
```
Please see comment in https://github.com/simphony/simphony-openfoam/pull/10/files#r27658560, There should be something that can be checked regarding the operation of `add_mesh` and `write_mesh` that idealy does not invlove `get_mesh`.

itziakos commented on foam_controlwrapper/tests/test_foam_interface.py:
```diff
@@ -0,0 +1,292 @@
+            puids = [self._foamPointLabelToUuid[lbl] for lbl in pointLabels]
+            self.assertEqual(puids, face.points)
+
+    def test_get_cell_points(self):
```
There should be a check where cells have Data since the foam_interface supports data attributes for cells correct?

khiltunen replied:
Data is not transferred through native interface so i think that it is enough to test the data in the cell in foam_mesh tests.

itziakos commented on foam_controlwrapper/tests/test_foam_interface.py:
```diff
@@ -0,0 +1,292 @@
+""" test foam native interface
```
Since these are tests for the foam_interface package they should be moved to the foam_interface package

itziakos commented on foam_controlwrapper/tests/test_foam_mesh.py:
```diff
@@ -0,0 +1,270 @@
+        cell.data[CUBA.VELOCITY] = (2, 1, 3)
+        foam_mesh.update_cell(cell)
+        cell_f = foam_mesh.get_cell(cell.uid)
+        self.assertEqual(cell.data[CUBA.VELOCITY], cell_f.data[CUBA.VELOCITY])
```
A better checck would be 
```python
self.assertIsInstance(cell_f.data, DataContainer)
self.assertEqual(cell_f.data, cell.data)
```

itziakos commented on foam_controlwrapper/tests/test_foam_mesh.py:
```diff
@@ -0,0 +1,270 @@
+        foam_mesh = FoamMesh('test_mesh', self.mesh)
+        for face_f in foam_mesh.iter_faces():
+            face = self.mesh.get_face(face_f.uid)
+            self.assertEqual(face.data[CUBA.LABEL],
```
I thought that `data` were only supported for *cells*, it looks that CUBA.LABEL is supported for face. But we cannot change them correct?

khiltunen replied:
CUBA.LABEL is used for grouping the faces for boundary conditions. We should have in the future sme grouping for mesh objects to get this done in a more elegant and efficient way. This information should also be saved along with the mesh itself.

itziakos commented:
@khiltunen, I had a  look in the PR and the new tests.  

The main comment is that a nubmer of test method just execute code without verify the behaviour which should be improved. In my opinion it would be nice for this issue to be fixed in this PR so that the tests are properly testing the implementation. However if you want to agument the tests later please add a new task issue in the issue tracker for fixing the offending test methods.

khiltunen commented:
I have now added verifying the code run in unittests where it was missing  @itziakos

itziakos commented on openfoam-interface/tests/test_foam_interface.py:
```diff
@@ -0,0 +1,316 @@
+        foamface.writeMesh(self.name)
+        meshpath = os.path.join(self.test_dir, self.name,
+                                'constant', 'polyMesh')
+        self.assertEqual(os.path.exists(
```
`suggestion`: It preferable to use `self.assertTrue` for checks that assert value == True (see table  https://docs.python.org/2/library/unittest.html#assert-methods), In general the preferable assert method is the one that is as specific as possible on what we aim to assert.

itziakos commented:
@khiltunen, I had another look, I do not have more comments at the moment, just a suggestion which you can fix if you want (there are always things that can be improved). It looks however, that some previous comments have been missed (e.g. https://github.com/simphony/simphony-openfoam/pull/10/files#r27653939)

Please update as necessary and If there are no comments from @nathanfranklin, I think that the PR will be ready merge.

khiltunen commented:
I looked through that missed previous comment and made corresponding changes. Also changed to tests to use assertionTrue.
